### DISCUSSION TITLE
Using libcurl with threaded resolver and CURLOPT_NOSIGNAL together on Linux

### DISCUSSION THREAD

**Discussion by SoulfreezerXP** [curl/curl#11324] [Q&A] [answered]
Maybe you can have a look at my stackoverflow-question here?
https://stackoverflow.com/questions/76488082/using-libcurl-with-threaded-resolver-and-curlopt-nosignal-together-on-linux

My question is whether it is RIGHT/OK to have the asynchronous THREADED_RESOLVER 
enabled and still use CURLOPT_NOSIGNAL = 1L ? Or will it even harm to use both together ?
(The threaded resolver is per default ENABLED, if I build with cmake and dont use c-ares, right?!)

I tried to study the libcurl-source-code (libcurl 7.72.0) and it seems, most of the signal-handling is anyway ignored, when build with the threaded-resolver ? But I am not sure and cannot see the whole picture, because it seems that in some cases of openssl a signal-stuff
could happen even when CURLRES_ASYNCH is defined (I am using OpenSSL)? 

This here are the parts I found in the source:

```
if(ENABLE_THREADED_RESOLVER)
  find_package(Threads REQUIRED)
  if(WIN32)
    set(USE_THREADS_WIN32 ON)
  else()
    **set(USE_THREADS_POSIX ${CMAKE_USE_PTHREADS_INIT})
    set(HAVE_PTHREAD_H ${CMAKE_USE_PTHREADS_INIT})**
  endif()
  set(CURL_LIBS ${CURL_LIBS} ${CMAKE_THREAD_LIBS_INIT})
endif()

#ifdef USE_ARES
#  define CURLRES_ASYNCH
#  define CURLRES_ARES
/* now undef the stock libc functions just to avoid them being used */
#  undef HAVE_GETADDRINFO
#  undef HAVE_FREEADDRINFO
#  undef HAVE_GETHOSTBYNAME
#elif defined(USE_THREADS_POSIX) || defined(USE_THREADS_WIN32)
**#  define CURLRES_ASYNCH
#  define CURLRES_THREADED**
#else
#  define CURLRES_SYNCH
#endif

#if defined(CURLRES_SYNCH) && \
    defined(HAVE_ALARM) && defined(SIGALRM) && defined(HAVE_SIGSETJMP)
/* alarm-based timeouts can only be used with all the dependencies satisfied */
#define USE_ALARM_TIMEOUT
#endif
```

The CURLOPT_NOSIGNAL-options is then set via:

```
case CURLOPT_NOSIGNAL:
    /*
     * The application asks not to set any signal() or alarm() handlers,
     * even when using a timeout.
     */
    data->set.no_signal = (0 != va_arg(param, long)) ? TRUE : FALSE;
break;
```

And set.no_signal is rarely used, because only if USE_ALARM_TIMEOUT would be set or 
USE_OPENSSL + HAVE_SIGACTION (?!)

The documentation is not really explicit about it, but it seems that it will work together and that 
CURLOPT_NOSIGNAL must be used in threaded applications **anyway**?
![grafik](https://github.com/curl/curl/assets/9033709/c5b372da-5dfc-4927-b98c-a0ad1339f5b0)

**Comment by bagder** [comment]
> The documentation is not really explicit about it

I don't understand what's not explicit. I think the screenshotted paragraphs are explicit.

> it seems that it will work together

Yes they work together, otherwise the documentation would say they don't.

> CURLOPT_NOSIGNAL must be used in threaded applications

I don't see this mentioned anywhere so I don't understand what makes you draw this conclusion.

You *can* use `CURLOPT_NOSIGNAL` when multithreaded but you don't have to. Most people don't. And if you do set it, you need to handle the outcome yourself. Like the risk that you get a SIGPIPE.

**Reply by SoulfreezerXP** [reply]
First of all thank you for your quick feedback!

I had read in some forums that people reported crashes when 
they did not have the CURLOPT_NOSIGNAL option set. This had really unsettled me, 
because I have changed an important application from single-threading 
to multi-threading. Now I am looking for the **safest** solution to not 
be surprised later by a crash.   

> I don't see this mentioned anywhere so I don't understand what makes you draw this conclusion.

The upper story and the following sentence from the docs makes me a bit uncertain which led me to the thought 
that you actually MUST have the option enabled to be on the safe side.

> "On some platforms, libcurl simply will not function properly multi-threaded unless the 
> CURLOPT_NOSIGNAL option is set." 

Maybe I have such a platform? (CentOS6 and CentOS7)

> "You can use CURLOPT_NOSIGNAL when multithreaded but you don't have to. 
> Most people don't."

Now I am a little confused. Am I understanding this correctly? So that would mean there are actually no 
dangers to omitting the flag (when threaded_resolver enabled)?

This is my scenario:
My application ignores SIGPIPES in general and always **before** a new thread is 
created even all signals are blocked (via sigmask) so that the new thread 
**inherits** the sigmask. The signal mask will be restored in the main thread afterwards.
In my signal handler I switch only one atomic bool, nothing more. 
In each thread I do a curl_easy_perform with an own handle.

I would appreciate to get a recommendation from you for the configuration 
for my scenario (flag set -> yes or no ?)

**Reply by bagder** [reply, marked as answer]
Using `CURLOPT_NOSIGNAL` is a *choice*. If the default behavior is not to your liking, then set this option and deal with the signals yourself.

Both ways are equally "safe". But if you don't want to deal with the signals in any other way than just ignoring them then I also don't see a reason why you can't just let libcurl do that by itself and just leave it at  default.

**Reply by SoulfreezerXP** [reply]
Thanks for your advice! I will try it without CURLOPT_NOSIGNAL for now.
